---
title: "batch effect correction, reclustering, and loading into loupe"
output: html_notebook
Author: Mike and Juan and Lisa
---

We see batch effects between Bright Field (BF) and Immunofluorescence (IF) imaging in Spatial Gene Expression data. Since data are from the same tissues, we would expect gene expression to be similar between sections before staining. Staining should not affect gene expression.

First, we combine Spatial Gene Expression data from two samples from the same tissue, each with a different staining protocol. The data can be combined by running the spaceranger aggr pipeline to combine these data using the following command:

```
spaceranger aggr --id=staining_IF_BF --csv=IF_BF.csv
```
When we load the .cloupe file output from the aggr pipeline into the Loupe Browser, we see this batch effect (see image below) where the two BF (blue) and IF (brown) samples are separated from each other.

Correcting for this batch effect should bring the cells into alignment.

There are a number of algorithms and tools for correcting batch effects. For more information on batch effects and batch effect correction, see this introduction: Batch Effect Correction.

For this tutorial we are going to use the Harmony batch effect correction algorithm (Korsunsky et al. 2019) implemented in the Seurat R package. The Harmony algorithm is available on GitHub, and the authors of Seurat wrote an integration function in the Seurat package.

##  Load Libraries

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(cowplot)
library(harmony)
```

# Download the files

The dataset with the IF filitered matrix can be found here:

https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Adult_Mouse_Brain

The dataset with the BF matrix can be found here:

https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Adult_Mouse_Brain_Coronal_Section_1


```{r}
setwd("~/")
# download IF filtered matrix:
download.file("https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Adult_Mouse_Brain_Coronal_Section_1/V1_Adult_Mouse_Brain_Coronal_Section_1_filtered_feature_bc_matrix.tar.gz","V1_Adult_Mouse_Brain_Coronal_Section_1_filtered_feature_bc_matrix.tar.gz")
# decompress the filtered matrix file just downloaded
untar("V1_Adult_Mouse_Brain_Coronal_Section_1_filtered_feature_bc_matrix.tar.gz")
file.rename("filtered_feature_bc_matrix","IF_brain_filtered_feature_bc_matrix")
# create a local directory called "BF_brain
download.file("https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Adult_Mouse_Brain/V1_Adult_Mouse_Brain_filtered_feature_bc_matrix.tar.gz","V1_Adult_Mouse_Brain_filtered_feature_bc_matrix.tar.gz")
untar("V1_Adult_Mouse_Brain_filtered_feature_bc_matrix.tar.gz")
file.rename("filtered_feature_bc_matrix","BF_brain_filtered_feature_bc_matrix")
```
## Load and Combine Data Sets

This section uses commands found in this Seurat vignette. It creates two Seurat objects and merges them together.

Seurat "objects" are a type of data that contain your UMI counts, barcodes, and gene features all in one variable. The Seurat tool has a function called "Read10X()" that will automatically take a directory containing the matrices output from Cell Ranger and input them into the R environment so you don't have to worry about doing this manually. We rely on this function to save us many steps.

We will start by creating two Seurat objects, one for each data set and then take a look at it:

```{r}
IF_brain.data <- Read10X(data.dir = "IF_brain_filtered_feature_bc_matrix/")
IF_brain<- CreateSeuratObject(counts = IF_brain.data, project = "IF")
IF_brain
BF_brain.data <- Read10X(data.dir = "BF_brain_filtered_feature_bc_matrix/")
BF_brain<- CreateSeuratObject(counts = BF_brain.data, project = "BF")
BF_brain
```

The output of "BF_brain" and "IF_brain" should be similar.

## Merge Objects

Use the merge function to combine the objects. Use the head and the tail functions to check how the data look, to make sure that they are equivalent.

```{r}
brain.combined <- merge(IF_brain, y = BF_brain, add.cell.ids = c("IF", "BF"), project = "2brains")
head(colnames(brain.combined))
tail(colnames(brain.combined))
```
This object creation and merging was adapted from the Vignette here: https://satijalab.org/seurat/v3.2/merge_vignette.html

## Visualize

First, visualize the data before running batch effect correction. Within this command, we will also be normalizing, scaling the data, and calculating gene and feature variance which will be used to run a PCA and UMAP.

```{r}
brain.combined <- NormalizeData(brain.combined) %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA(verbose = FALSE)
brain.combined <- RunUMAP(brain.combined, dims = 1:30)
DimPlot(brain.combined, group.by = "orig.ident")
```
We can see the two samples are not mixed well. But they are better than they were before (see Loupe above). This is probably an effect of normalization, which can help sometimes to correct for small differences.

## Run harmony
The next few commands are adapted from the vignette here: https://htmlpreview.github.io/?https://github.com/satijalab/seurat.wrappers/blob/master/docs/harmony.html
```{r}
#brain.combined <- NormalizeData(brain.combined) %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA(verbose = FALSE)
brain.combined <- RunHarmony(brain.combined, group.by.vars = "orig.ident")
brain.combined <- RunUMAP(brain.combined, reduction = "harmony", dims = 1:30)
brain.combined <- FindNeighbors(brain.combined, reduction = "harmony", dims = 1:30) %>% FindClusters()
DimPlot(brain.combined, group.by = "orig.ident")
DimPlot(brain.combined, group.by = "ident")
DimPlot(brain.combined, group.by = "ident", split.by = 'orig.ident')
```
Now the two samples are mixed nicely together.

# Export the UMAP projection to a csv that can be loaded into Loupe
Some of the code below is from: https://github.com/satijalab/seurat/issues/1391
Split the matrix, edit the barcode into a Loupe-compatible format, then merge into a new matrix.
```{r}
corrected.data <- SplitObject(brain.combined, split.by = "orig.ident")

IF.barcode <- rownames(Embeddings(object = corrected.data$IF, reduction = "umap"))
IF.barcode <- gsub("IF_","",IF.barcode)
IF.barcode <- gsub('.{2}$','',IF.barcode)
IF.barcode <- paste(IF.barcode,"-1", sep="")
IF.proj <- Embeddings(object = corrected.data$IF, reduction = "umap")
UMAP.IF <- cbind("Barcode" = IF.barcode, IF.proj)

BF.barcode <- rownames(Embeddings(object = corrected.data$BF, reduction = "umap"))
BF.barcode <- gsub("BF_","",BF.barcode)
BF.barcode <- gsub('.{2}$','',BF.barcode)
BF.barcode <- paste(BF.barcode,"-2", sep="")
BF.proj<- Embeddings(object = corrected.data$BF, reduction = "umap")
UMAP.BF <- cbind("Barcode" = BF.barcode, BF.proj)

corrected.umap <- rbind(UMAP.IF,UMAP.BF)
write.table(corrected.umap, file="corrected_umap.csv", sep = ",", quote = F, row.names = F, col.names = T)
```

## Export the clusters to a csv that can be loaded into Loupe
```{r}
clusters.IF = Idents(corrected.data$IF)
clusters.IF.data <- cbind("Barcode" = IF.barcode, data.frame("clusters" = clusters.IF))

clusters.BF = Idents(corrected.data$BF)
clusters.BF.data <- cbind("Barcode" = BF.barcode, data.frame("clusters" = clusters.BF))

corrected.cluster <- rbind(clusters.IF.data, clusters.BF.data)
write.table(corrected.cluster, file="corrected_clusters.csv", sep = ",", quote = F, row.names = F, col.names = T)
```

